src/base.js

```diff
@@ -323,7 +323,84 @@ class Base {
   /**
    * Returns a promise
    */
-  handleThirdPartyRoomMessage(thirdPartyRoomMessageData, doNotTryToGetRemoteUserStoreData) {
+  getUserClient(roomId, senderId, senderName, avatarUrl, text, doNotTryToGetRemoteUserStoreData) {
+    const { info } = debug(this.getUserClient.name);
+    info("get user intent for third party user %s (%s)", senderId, senderName);
+
+    if (senderId === undefined) {
+      return Promise.resolve(this.puppet.getClient());
+    } else {
+      if (!senderName && !this.allowNullSenderName) {
+        if (doNotTryToGetRemoteUserStoreData)
+          throw new Error('preventing an endless loop');
+
+        info("no senderName provided with payload, will check store");
+        return this.getOrInitRemoteUserStoreDataFromThirdPartyUserId(senderId).then((remoteUser)=>{
+          info("got remote user from store, with a possible client API call in there somewhere", remoteUser);
+          info("will retry now");
+          const senderName = remoteUser.get('senderName');
+          return this.getUserClient(roomId, senderId, senderName, true);
+        });
+      }
+
+      info("this message was not sent by me");
+      const ghostIntent = this.getIntentFromThirdPartySenderId(senderId);
+      let promiseList = [];
+      promiseList.push(() => ghostIntent.join(roomId));
+
+      if (senderName)
+        promiseList.push(() => ghostIntent.setDisplayName(senderName));
+
+      if (avatarUrl)
+        promiseList.push(() => this.setGhostAvatar(ghostIntent, avatarUrl));
+
+      return Promise.all(promiseList).then(() => ghostIntent.getClient());
```

Comment:
Ah that's why this `mapSeries` was here! I must confess I don't really understand the `p => p()`: we execute each promise?

I think I prefer the "`join` first then `Promise.all`" way.